src/chkstat.c

```diff
@@ -1223,22 +1223,28 @@ main(int argc, char **argv)
           fprintf(stderr, "%s: chmod: %s\n", e->file+rootl, strerror(errno));
           errors++;
         }
-      if (!caps_ok)
+      // chown and - depending on the file system - chmod clear existing capabilities
+      // so apply the intended caps even if they were correct previously
+      if (e->caps || !caps_ok)
         {
           if (S_ISREG(stb.st_mode))
             {
               // cap_set_file() tries to be helpful and does a lstat() to check that it isn't called on
               // a symlink. So we have to open() it (without O_PATH) and use cap_set_fd().
-              int cap_fd = open(fd_path, O_NOATIME | O_CLOEXEC);
+              int cap_fd = open(fd_path, O_NOATIME | O_CLOEXEC | O_RDONLY);
```

Comment:
Is a O_RDONLY file descriptor really good for changing capabilities of the file? Couldn't find any definitive documentation for that quickly ....

Reply:
I've tested (on ext4), and it worked. The documentation of the underlying syscalls `fsetxattr`/`fremovexattr` indeed doesn't say anything either way.

Investigating this, [1] lead me to [2] - the kernel doesn't care. It takes the identical code paths for fd-based and path-based calls, once they are translated to a `dentry`.

1: https://lists.gnu.org/archive/html/bug-coreutils/2009-09/msg00166.html
2: https://elixir.bootlin.com/linux/v5.5/source/fs/xattr.c#L495

Reply:
Good to know. Then I think this is fine the way it is, even if the semantics are strange.